chef-config/lib/chef-config/workstation_config_loader.rb

```diff
@@ -167,7 +167,11 @@ def apply_credentials(creds, profile)
         when "client_key"
           extract_key(value, :client_key, :client_key_contents)
         when "knife"
-          Config.knife.merge!(value.transform_keys(&:to_sym))
+          value = value.transform_keys(&:to_sym)
+          if value[:secret_file]
```

Comment:
@drewhammond So, expand_path with a default relative/base path of `home_chef_dir`? It might be worth a comment on the logic that's followed here.

Reply:
Yeah, this is just borrowed from the existing `extract_key()` method https://github.com/chef/chef/blob/a6853a4a54e3ac0a4cf37e4c778aa59e237480f2/chef-config/lib/chef-config/workstation_config_loader.rb#L186 which probably could have been reused here too, but it seemed like a bad/forced fit (`secret_file` will never contain the raw key contents).

Reply:
AFAIK, these are the only three properties that can contain path-like values. If there are others then I think it makes sense to break this out as a new `abs_or_relative_path()` method and use it for all, but since `validator_key` and `client_key` can also contain a raw key instead of a path, it makes it a bit awkward.

What do you think?

Reply:
Ok, didn't want it to block, but definitely wanted to ask the question.

Reply:
It looks to be a bit messy if wanted to fully implement it for all potential path operations. According to https://docs.chef.io/workstation/knife_setup/#setting-up-knife the credentials file can take any key=value pair found in the config.rb https://docs.chef.io/workstation/config_rb/ and that file is also extended to take any knife config options as optional config settings https://docs.chef.io/workstation/config_rb_optional_settings/. Essentially would need to go through all of those and list them out to check for expanding the path.

Reply:
Yeah, I think I'm afraid of biting off more than I can chew with that. For completeness, here's all the settings that accept some form of path

- `bootstrap_template`
- `client_d_dir`
- `client_key`
- `cookbook_path` (string or array)
- `validation_key`
- `config_log_location` (string, symbols)
- `knife[:bootstrap_template]`
- `knife[:bootstrap_vault_file]`
- `knife[:chef_repo_path]`
- `knife[:config_file]`
- `knife[:cookbook_path]`
- `knife[:download_directory]`
- `knife[:file]`
- `knife[:identity_file]`
- `knife[:repository]`
- `knife[:script_path]`
- `knife[:secret_file]`
- `knife[:template_file]`
- `knife[:user_home]`

Agreed, it'd be messy. Save it for a rainy day?

Reply:
I think we're all in agreement that this needs a more generic solution.

If for no other reason than the user experience for secret_file working, but them something else _not_ working is way worse than always having to specify `~/....`.